# Sequencing and data quality tutorial

This tutorial will guide you through the essential steps to evaluate and preprocess sequencing data, focusing on FASTQ file handling, quality assessment using **FastQC**, and data trimming with **Fastp**. These steps are critical for ensuring high-quality inputs for downstream analyses in conservation genomics.

## Section 1: Understanding Sequencing Output ‚Äì FASTQ Files

1.1) Understand FASTQ files

FASTQ files are the standard format for storing raw sequencing data. Each read in a FASTQ file consists of four lines:

1.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Sequence Identifier**: Starts with @ and includes metadata.

2.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Nucleotide Sequence**: The sequence of DNA bases (A, T, G, C).

3.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Separator Line**: Starts with + and may repeat the identifier.

4.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Quality Scores**: ASCII-encoded quality values corresponding to each base in the sequence.

![image](https://github.com/user-attachments/assets/5ffbd3ff-adb8-4fb0-81bc-4c66609771bd)

1.2) Understanding base quality (PHRED score) and the ASCII code

Each base in the FASTQ file has an associated quality score Q, which reflects the probability p that the base is incorrect. The formula to get Q is:

![image](https://github.com/user-attachments/assets/e8f0e2b2-5112-4ebc-888c-3126d8886ec0)

The values of Q can range from 0 to 93 (but the maximum score you will usually see is 40). To save space in a FASTQ file, each quality score is transformed into a single ASCII character.

![image](https://github.com/user-attachments/assets/5d0a8f1b-4a1d-4777-b8c9-a4325d242560)

üìù **Exercise:**

Use basic UNIX commands to explore the data and get yourselves familiar with the FASTQ format.

1) Use less to inspect a fastq file. What are the initial header characters for all reads?

```bash
zless /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R1.fastq.gz
```

2) How many reads do we have for each individual? Use the command line below to count the lines (`zgrep -c`) that start with (`^`) the specified header characters (`@SR`).

```
zgrep -c "^@SR" /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R1.fastq.gz
```

3) What are the read lengths? This `awk` command line processes a compressed FASTQ file to analyze the distribution of read lengths. It first decompresses the file using zcat, then uses awk to extract the sequence lines (every 4th line, starting with the second). For each sequence line, it calculates the length of the sequence. The resulting lengths are then sorted and counted using sort and uniq -c, respectively, providing a summary of how many reads have each specific length.

```
zcat /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R1.fastq.gz | awk '(NR%4==2) {print length($0)}' |
 sort | uniq -c
```

4) Getting familiar with quality values:
Use the ASCII table below to get the numerical value for the ASCII character *.

![image](https://github.com/user-attachments/assets/eb398136-18e2-44b7-a4d1-74d2b9de937f)

It is 42. The convention is to subtract 33 (phred33 encoded), which makes 9. Is this a good quality? To find out the p value, calculate:

![Screenshot 2025-01-05 121936](https://github.com/user-attachments/assets/8df4b8c6-c639-49fb-a28b-39016cac6158)

Now that you know how to convert quality values fill out the following table:

|Quality in fastq |‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇQ in decimal |‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇp|
|---|---|---|
| *| 9|  0.1259|
| I |‚ÄÇ40‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ|  |


## Section 2: Evaluating Data Quality with FastQC

FastQC is a widely used tool for evaluating sequencing data quality. It generates a detailed report that includes metrics like:

- **Per Base Quality**: Average quality scores across the length of reads.
- **GC Content**: The percentage of G and C bases.
- **Duplicate Levels**: Identifies potential PCR duplicates.
- **Adapter Content**: Checks for residual sequencing adapters.


üìù **Exercise:**

We will submit a job for running FastQC for two fastq files: read 1 and read 2. Reminders:
- Go to the directory `02_sequencing_dataquality`
- Use `vim` or `nano` to create a script file named <jobname>.sh
- Replace 'YOUR_USERNAME' with your own username
- Submit the job using: `sbatch <jobname>.sh`
- Check the job status with: `squeue -u <your_username>`

```bash
#!/bin/bash
#SBATCH --job-name fastqc
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/logs/fastqc.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/logs/fastqc.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

# 1) Load FastQC using a module:
module load fastqc

# 2) Run FastQC on read 1 and 2 files:
fastqc /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R1.fastq.gz -o /anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality
fastqc /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R2.fastq.gz -o /anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality
```

- Navigate to the `/anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality` directory in the browser, download the HTML file and open it in a browser to examine the quality metrics.

> [!IMPORTANT]
> :grey_question: Evaluate the quality of the read 1 and 2 fastq files. How long are the reads 1 and 2 on average? How is the quality of the data? Are there any drops in quality along the sequence? Which file is of better quality? Are there any adapter sequences detected? A good explanation of the different plots can be found at [here](https://rtsf.natsci.msu.edu/genomics/technical-documents/fastqc-tutorial-and-faq.aspx?utm_source=chatgpt.com)

## Section 3: Trimming and Filtering with Fastp

Fastp is an efficient tool for quality control and preprocessing. It can:

- Trim low-quality bases.
- Remove adapters.
- Filter short reads.
- Generate QC reports.

üìù **Exercise:**

1) Let's use the following script to run Fastp to trim adapter sequences and filter paired-end reads based on quality. The tool automatically detects adapter sequences and removes them. We set `-q` to 20, which ensures that only reads with a Phred quality score of at least 20 are retained. Make sure to update the placeholder <YOUR_USERNAME> with your username before running the script.
   
```bash
#!/bin/bash
#SBATCH --job-name fastp
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/logs/fastp.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/logs/fastp.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

# 1) Load Fastp on your HPC:
module load biocontainers/default
module load fastp

# 2) Run Fastp to trim paired-end reads:
fastp -i /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R1.fastq.gz -I /anvil/projects/x-bio240351/shared_data/fastq/SRR957824_500K_R2.fastq.gz -o /anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality/SRR957824_500K_R1_trimmed.fastq -O /anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality/SRR957824_500K_R2_trimmed.fastq --html fastp_report.html -q 20
```

Note that the key parameters in Fastp that we are using are:

- `-i`: read1 input file name
- `-I`: read2 input file name
- `-o`: read1 output file name
- `-O`: read2 output file name
- `--html`: the html format report file name
- `-q` or `--qualified_quality_phred`: the quality value that a base is qualified.

> [!IMPORTANT]
> :grey_question: Check the HTML report generated by Fastp and answer the following questions: What was the number of reads before and after filtering? What percentage of adapters was detected and filtered out? What was the percentage of reads with low quality, reads containing too many N bases, and reads that were too short?

2) If you have time, compare pre- and post-trimming quality rerunning FastQC on the trimmed data to confirm improvements:

```bash
#!/bin/bash
#SBATCH --job-name fastqc_postrimming
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/logs/fastqc_postrimming.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/logs/fastqc_postrimming.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

# 1) Load FastQC using a module:
module load fastqc

# 2) Run FastQC on the trimmed files:
fastqc /anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality/*trimmed.fastq -o /anvil/scratch/YOUR_USERNAME/02_sequencing_dataquality
```

> [!IMPORTANT]
> :grey_question: Compare the pre- and post-trimming reports to evaluate changes in: Base quality scores, Adapter content, Read length distribution.

## üí° Tips: Best Practices for Data Quality

1.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Retain Metadata**: Keep logs and summaries from FastQC and Fastp for reproducibility.
2.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Optimize Parameters**: Tailor Fastp settings (e.g., trimming thresholds) based on FastQC reports.
3.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Batch Processing**: Use loops or job arrays for large datasets.
4.‚ÄÇ‚ÄÇ‚ÄÇ‚ÄÇ**Save Outputs**: Maintain a structured directory with raw, trimmed, and quality-checked data.


By following this workflow, you ensure your sequencing data is high-quality and ready for downstream analysis. These tools and techniques are foundational in conservation genomics, helping to maximize the reliability and accuracy of your results.

