# Read Mapping and Variant Calling 

## Part one: Read Mapping

### Quick review of file types for Read Mapping

- fastq files -> this file stores raw sequence data from the sequencing machine. For paired-read sequencing, wew ill obtain a Read 1 (R1) file and Read 1 (R2) file.  

Let's look at our fastq file: 

```
less /anvil/projects/x-bio240351/shared_data/fastq/CH0893_ACCTTGGC-ATGAGGCC_L001_R1_001.fastq.gz
```

The four lines correspond to: 
```
1. A sequence identifier with information about the sequencing run and the cluster. 

2. The sequence (the base calls; A, C, T, G and N).

3. A separator, which is simply a plus (+) sign.

4. The base call quality scores. These are Phred +33 encoded, using ASCII characters to represent the numerical quality scores.
```

- Let's look at our BAM file, which stands for Binary Alignment Map:
```
module load samtools 

samtools view /anvil/projects/x-bio240351/shared_data/bams/NA5206.loxAfr4_NC000934_subsampled.bam | less -S
```

<img width="746" alt="image" src="https://github.com/user-attachments/assets/b01666bf-fa37-4a7b-a48a-4507a56be965" />

- Mapping quality -> probability that the read is misaligned in the BAM file. Reported from 0 to 255, with a lower number higher quality. 
    - MAPQ score 20, then the probability of a correctly mapped read was 0.99
    - MAPQ score 30, then the probability of a correctly mapped read was 0.999
- CIGAR score -> compressed representation of the alignment (Compact Idiosyncratic Gapped Alignment Report)


###  Alignment of reads to reference genome

First, we need to align our trimmed fastq files to the reference genome. 

> [!IMPORTANT]
> What is a reference genome and where did we get it? We can search for previously published reference genomes on the National Center for Biotechnology Information (NCBI)'s website (https://www.ncbi.nlm.nih.gov/search/)

- Click here: https://www.ncbi.nlm.nih.gov/search/
  
-  Type in African elephant into the search
  
-  How many African elephant reference genomes are there?
  
-  https://www.ncbi.nlm.nih.gov/datasets/genome/?taxon=9784

First, we need to make a job file that contains the script we want to run. 

```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling
touch 01_alignment.sh
```
Then we will copy and paste this information below into our alignment script: 

```
vim 01_alignment.sh
#paste the contents below
:wq
```

Script for alignment
```
#!/bin/bash
#SBATCH --job-name 01_alignment
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/01_alignment.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/01_alignment.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

#load our modules

module load bwa
module load samtools

#We can list our files here, then call them later in the script

R1=/anvil/projects/x-bio240351/shared_data/fastq/NA5206_Namibia_R1.fastq
R2=/anvil/projects/x-bio240351/shared_data/fastq/NA5206_Namibia_R2.fastq
REF=/anvil/projects/x-bio240351/shared_data/reference/Chromosomes.v2.fasta
OUTDIR=/anvil/scratch/x-YOURNAME/03_readmapping_variantcalling

#bwa index ${REF} #index the reference genome of the Elephant

bwa mem -R "@RG\tID:NA5206\tSM:NA5206" ${REF} ${R1} ${R2} | \
samtools view -q 10 -h -b - | \
samtools sort -o ${OUTDIR}/NamibiaElephant_NA5206.bam -
```
- bwa index -> creates an index to make searching for the reference genome much faster
- bwa mem -> the alignment algorithm for aligning sequence reads based on sequence simularity
- -R -> read group to give a unique tag based on sequence origin
- $REF -> reference genome
- samtools view -q 10 -> keeping read with at least a MAPQ of 10
- samtools sort -> sort our bam file by chromosome position
- 
Then we can run the script: 
```
sbatch 01_alignment.sh
```

We can view the out file to see how the job is running: 

```
squeue -u x-YOURNAME
less /anvil/scratch/YOUR_USERNAME/log/01_alignment.e*
```

Let's check to see if the bam file is here and starting to be formed: 
```
module load samtools
samtools view /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/NamibiaElephant_NA5206.bam | less -S 
```
We can see the job is running. This will take a long time to align the fastq file to the reference genome, so we will use our previously made bam file. 

### Depth of our bam file

We will calculate the depth of our bam file. The **depth** is the average number of reads that align to a specific position on the reference genome. **Coverage** refers to the percentage of the genome that is coveraged by at least one reads. 

First, we need to create a job file to place our script into: 
```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling
touch 02_depth.sh
```

Then we will copy and paste this information below into our alignment script: 

```
vim 02_depth.sh
#paste the contents below
wq
```
Script for calculating depth for a bam file: 
```
#!/bin/bash
#SBATCH --job-name 02_depth
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/02_depth.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/02_depth.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load samtools

BAM=/anvil/projects/x-bio240351/shared_data/bams/NA5206.loxAfr4_NC000934_subsampled.bam

samtools depth ${BAM}  |  awk '{sum+=$3} END { print "Average = ",sum/NR}' > depth_NA5206.txt
```
Then we can run the script: 
```
sbatch  02_depth.sh

```
Now we can see a new file in our folder called `depth_NA5206.txt` by doing `ls`

```
less /anvil/projects/x-bio240351/shared_data/outputs/depth_NA5206.txt
```
> [!IMPORTANT]
> What is the depth of our bam file? 

We can also calculate depth per position in our bam file: 

First, we need to create a job file to place our script into: 
```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling
touch 02_perpositiondepth.sh
```

Then we will copy and paste this information below into our alignment script: 

```
vim 02_perpositiondepth.sh
#paste the contents below
wq
```
Script for calculating depth for a bam file: 
```
#!/bin/bash
#SBATCH --job-name 02_depthperposition
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-lhennelly/log/02_depthperposition.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/x-lhennelly/log/02_depthperposition.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load samtools

BAM=/anvil/projects/x-bio240351/shared_data/bams/CH0893.loxAfr4_NC000934.bam

samtools depth ${BAM} > depth_NA5206_per_position.txt
```
Then we can run the script: 
```
sbatch  02_perpositiondepth.sh

```
Now we can see a new file in our folder called `depth_NA5206.txt` by doing `ls`

```
less 02_perpositiondepth.txt
```

### Visualizing read alignment 
There are a few programs where we can visualize read alignments, some including the Integrative Genome Viewer and tview in samtools. 

We will visualize some reads in the alignment using IGV (https://igv.org/doc/desktop/#DownloadPage/). If you want, download the IGV software so we can visualize some reads. 

Download the "wSierraMorena.CanFam31.realigned_TSHR.bam" and "wSierraMorena.CanFam31.realigned_TSHR.bam.bai" files onto your computer. 

Change the reference genome to "Canfam3.1" and upload the bam file under File -> Load File. Change the genomic coordinate location to be: "chr8:53,400,332-53,416,468". 

Now you can visualize the reads, quality of the reads, and where there are split reads and mate pairs. 

> [IMPORTANT!]
> What do we observe in the region "chr8:53,400,332-53,416,468"? What could explain the poor quality in mapping? 

### Calculating read flags 

We can use samtools flagstat to count the  number of alignments for each FLAG type. 

The first row gives an output of total number of reads that are QC pass and fail. 

First, we need to create a job file to place our script into: 
```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling
touch 02_flatstat.sh
```

Then we will copy and paste this information below into our alignment script: 

```
vim 02_flatstat.sh
#paste the contents below
wq
```
Script for calculating depth for a bam file: 
```
#!/bin/bash
#SBATCH --job-name 02_flatstat
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-lhennelly/log/02_flatstat.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/x-lhennelly/log/02_flatstat.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load samtools

BAM=/anvil/projects/x-bio240351/shared_data/bams/CH0893.loxAfr4_NC000934.bam

samtools flagstat ${BAM} > NA5206_flagstat.txt
```
Then we can run the script: 
```
sbatch  02_flatstat.sh

```
Now we can see a new file in our folder called `NA5206_flagstat.txt` by doing `ls`

```
less NA5206_flagstat.txt
```


## Part two: Variant Calling with GATK


The GATK (Genome Analysis Toolkit) is one of the most used programs for genotype calling in sequencing data in model and non model organisms. However, the GATK was designed to analyze human genetic data and all its pipelines are optimized for this purpose (https://gatk.broadinstitute.org/hc/en-us).

### Quick background on file types: 


- A fastq.gz file contains the raw reads that are not aligned to the reference genome. Let's see what a fastq.gz file looks like: 
```
less /anvil/projects/x-bio240351/shared_data/fastq/NA5206_AGTTGGCT-GCAACCAT_L003_R2_001.fastq.gz

```
- A bam file is called a Binary Alignment Map and it is a file where the reads of an individual are aligned to a reference genome. So we know where each read goes on the genome. 
```
samtools view /anvil/projects/x-bio240351/shared_data/bams/NA5206.loxAfr4_NC000934_subsampled.bam | less -S

```

- A variant call format (VCF) file contains the variants along the genome for a single individual, or multiple individuals. 
```
zless -S /anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_imputed_sites_variable_nomissing_maf005_no1stdegree.vcf.gz
```


![image](https://github.com/user-attachments/assets/fd196f2e-4bdb-4a65-a558-98844d023c1f)

### FORMAT column information: 

The format column lists fields separated by colons. For example, GT:GQ. The list of fields provided depends on the variant caller used. Available fields include:

•	AD—Allele Depth; if the GT is 0/0, the AD is the reference count. If the GT is 0/1 or 1/1, the AD is of the form X,Y, where X is the reference allele count and Y is the alternative allele count. If the GT is 1/2, the AD is of the form Y,Z, where Y and Z are the alternative allele 1 and 2 counts.
•	DP—Total depth used for variant calling.
•	GQ—Genotype quality.
•	GT—Genotype. 0 corresponds to the reference base, 1 corresponds to the first entry in the ALT column, and so on. The forward slash (/) indicates that no phasing information is available.
•	NL—Noise level; an estimate of base calling noise at this position.
•	SB—Strand bias at this position. Larger negative values indicate less bias; values near 0 indicate more bias.
•	VF—Variant frequency; if the GT is 0/0, the VF is the nonreference frequency. If the GT is 0/1 or 1/1, the VF is the frequency of the variant allele. If the GT is 1/2, the VF is the frequency of the two variant alleles, together.


### HaplotypeCaller
HaplotypeCaller calls SNPs and indels via local de-novo assembly of haplotypes. This means when it encounters a region showing signs of variation, it discards mapping information and reassembles the reads in that region. This just allows HaplotypeCaller to be more accurate wen calling regions that are typically difficult to call. 

First, we will make a new script for Haplotypecaller 
```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling
touch 03_GATK_HaplotypeCaller.sh
```

Then we can run the script:

```
#!/bin/bash
#SBATCH --job-name haplotypecaller
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-YOUR_USERNAME/log/haplotypecaller.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/x-YOUR_USERNAME/log/haplotypecaller.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load gatk/4.1.8.1

BAM=/anvil/projects/x-bio240351/shared_data/bams/NA5206.loxAfr4_NC000934_subsampled.bam

REF=/anvil/projects/x-bio240351/shared_data/reference/Chromosomes.v2.fasta

gatk HaplotypeCaller \
-I ${BAM} \
-R ${REF} \
-ERC GVCF \
-O /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/NA5206.loxAfr4_NC000934_subsampled.g.vcf

```

Let's sbatch the script and look at our output file to see the job running: 
```
sbatch 03_GATK_HaplotypeCaller.sh
```
Check your squeue to see if you're job is running: 
```
squeue -u x-YOURNAME
```
And next we'll check the error file to see how the script is running: 
```
less /anvil/scratch/YOUR_USERNAME/log/03_GATK_HaplotypeCaller.o%j
```
We can see the job is running, and if we refresh looking at `/anvil/scratch/YOUR_USERNAME/log/03_GATK_HaplotypeCaller.o%j` we can see it is computationally working. 

Let's look at the G.VCF file: 
```
less /anvil/projects/x-bio240351/shared_data/gvcf/NA5206.loxAfr4_NC000934_subsampled.gvcf
```
> [!IMPORTANT]
> How is this different than a multi-sample VCF file? What does the <NON_REF> mean?
> What is the depth of the chr 1 position 3000055?
> Can you find a called variant that is different than the reference genome call? 


### GenomeDBImport
This step creates a GenomicsDB datastore to speedup joint genotyping. This in prep for the Genotype calling.

First, we will make a script: 
```
cd /anvil/scratch/x-YOUR_USERNAME/03_readmapping_variantcalling
touch 04_GATK_GenomeDBImport.sh
```

There's a few things we need to do in order to prepare for GenomeImport. We need to make a sample map: 
```
less /anvil/projects/x-bio240351/shared_data/bams/samplemap.txt
```

And we will make a directory for scratch and output of GenomeImport within our directory
```
mkdir /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenomeImport_scratch
mkdir /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenomeImport
```

Then we will paste this script below: 
```
#!/bin/bash
#SBATCH --job-name genomeImport
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/genomeImport.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/genomeImport.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load gatk/4.1.8.1

OUTDIR=/anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenomeImport
TEMPDIR=/anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenomeImport_scratch
SAMPLEMAP=/anvil/projects/x-bio240351/shared_data/bams/samplemap.txt

gatk GenomicsDBImport --genomicsdb-workspace-path ${OUTDIR}/elephant_gvcf_db --batch-size 50 -L chr1 --sample-name-map ${SAMPLEMAP} --tmp-dir ${TEMPDIR}
```

And we can run the file: 
```
sbatch 04_GATK_GenomeDBImport.sh
```

Then lets look at our Genomeimport directory: 
```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenomeImport
```
And our output file: 
```
/anvil/scratch/x-YOUR_USERNAME/log/genomeImport.e%j 
```

If we cd into your GenomeImport directories, we will see it worked: 
```
cd /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenomeImport
ls
```

### GenotypeGVCFs
Now we can perform joint genotyping that were pre-called with HaplotypeCaller.

- It will look at the available information for each site from both variant and non-variant alleles across all samples, and then produce a multi-sample VCF that contains only sites that are found as variant in at least one sample.

https://gatk.broadinstitute.org/hc/en-us/articles/360037057852-GenotypeGVCFs

Let's make a directory for our VCF output:
```
mkdir /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenotypeGVCF
mkdir /anvil/scratch/x-YOURNAME/03_readmapping_variantcalling/GenotypeGVCF_scratch
```
Now we can make a job file and paste our script into it: 
```
touch 05_GATK_GenotypeGVCFs.sh
```

Now let's run the job file:
```
#!/bin/bash
#SBATCH --job-name genotypecaller
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-lhennelly/log/05_GATK_GenotypeGVCFs.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/x-lhennelly/log/05_GATK_GenotypeGVCFs.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load gatk/4.1.8.1

REF=/anvil/projects/x-bio240351/shared_data/reference/Chromosomes.v2.fasta
OUTDIR=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/GenotypeGVCF
TEMPDIR=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/GenotypeGVCF_scratch

cd /anvil/scratch/x-lhennelly/03_readmapping_variantcalling/GenomeImport

gatk GenotypeGVCFs -R ${REF} -V gendb://elephant_gvcf_db -O ${OUTDIR}/gatk_Elephants.vcf.gz --tmp-dir ${TEMPDIR}
```
And we can look at our results: 
```
cd /anvil/scratch/x-lhennelly/03_readmapping_variantcalling/GenotypeGVCF
less gatk_Elephants.vcf.gz
```

Let's look at the multi-sample VCF we will use for the course, with all elephant samples: 
```
zless -S /anvil/projects/x-bio240351/shared_data/vcf/elephants.chr.11.12.subset.vcf.gz
```

## Variant Filteration 
Now we will filter our variants based on certain criteria.

- filter "DP < 1800" -> remove sites that have very high depth (why is this important?)

- filter FS -> FisherStrand, accounts for strand bias at the site. Strand bias is whether the alternate allele was seen more or less often on the forward or reverse strand than the reference allele. This results in the incorrect amount of evidence

- List of rest of filter criteria: https://gatk.broadinstitute.org/hc/en-us/articles/360037434691-VariantFiltration
  
Now we can make a job file and paste our script into it: 
```
touch 06_GATK_VariantFilteration.sh
```

Now let's run the job file:
```
#!/bin/bash
#SBATCH --job-name VariantFilteration
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-USERNAME/log/06_GATK_VariantFilteration.o%j     # Name of stdout output file
#SBATCH -e /anvil/scratch/x-USERNAME/log/06_GATK_VariantFilteration.e%j     # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load gatk/4.1.8.1

VCF=/anvil/projects/x-bio240351/shared_data/vcf/elephants.chr.11.12.subset.vcf
OUT=/anvil/scratch/x-USERNAME/03_readmapping_variantcalling/elephants.chr.11.12.subset_filtered.vcf

gatk VariantFiltration -V ${VCF} -filter "DP < 31800" --filter-name "DP1800" -filter "SOR > 3.0" --filter-name "SOR3" -filter "FS > 60.0" --filter-name "FS60" -filter "MQ < 40.0" --filter-name "MQ40" -filter "MQRankSum < -12.5"  --filter-name "MQRankSum-12.5" -filter "ReadPosRankSum < -8.0" --filter-name "ReadPosRankSum-8" -O ${OUT}
```
> [!IMPORTANT]
> filter "DP < 1800" -> remove sites that have very high depth
> Why is this important to remove very high depth?

- filter FS -> FisherStrand, accounts for strand bias at the site. Strand bias is whether the alternate allele was seen more or less often on the forward or reverse strand than the reference allele. This results in the incorrect amount of evidence


Then let's sbatch this: 
```
sbatch 06_GATK_VariantFilteration.sh
```
We can check if our job is running by: 
```
squeue -u x-USERNAME
```

### Filtering with VCFtools 

VCFtools is an excel

```
touch 07_VCFtoolsfiltering.sh
```
Copy and paste this script
```
#!/bin/bash
#SBATCH --job-name VCFtoolsfiltering
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-lhennelly/log/07_VCFtoolsfiltering.o     # Name of stdout output file
#SBATCH -e /anvil/scratch/x-lhennelly/log/07_VCFtoolsfiltering.e     # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

head -n 5000 /anvil/projects/x-bio240351/shared_data/vcf/elephants.chr.11.12.subset.vcf > /anvil/scratch/x-lhennelly/03_readmapping_variantcalling/elephants.chr.11.12.subset_vcftools.vcf
```
Run the script: 
```
sbatch 07_VCFtoolsfiltering.sh
```

Now we will calculate the site quality and plot: 
```
touch 08_VCFtoolsfiltering_sitequality.sh

```
Copy and paste this: 
```
#!/bin/bash
#SBATCH --job-name VCFtoolsfiltering_sitequality
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-lhennelly/log/08_VCFtoolsfiltering_sitequality.o     # Name of stdout output file
#SBATCH -e /anvil/scratch/x-lhennelly/log/08_VCFtoolsfiltering_sitequality.e     # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load vcftools

VCF=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/elephants.chr.11.12.subset_vcftools.vcf
OUT=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/elephants.chr.11.12.subset_vcftools_sitequality
OUT_depth=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/elephants.chr.11.12.subset_vcftools_depth
OUT_missingsite=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/elephants.chr.11.12.subset_vcftools_missingsite
OUT_allelefreq=/anvil/scratch/x-lhennelly/03_readmapping_variantcalling/elephants.chr.11.12.subset_vcftools_allelefreq

vcftools --vcf ${VCF} --site-quality --out ${OUT}
vcftools --vcf ${VCF} --site-mean-depth --out {$OUT_depth}
vcftools --vcf ${VCF} --missing-site --out ${OUT_missingsite}
vcftools --vcf ${VCF} --freq2 --out ${OUT_allelefreq} --max-alleles 2
```
Let's look at the output: 
```
less elephants.chr.11.12.subset_vcftools_sitequality.lqual
```
Now let's plot in R 

```
R
library(tidyverse)
library(ggplot2)
var_qual <- read_delim("./elephants.chr.11.12.subset_vcftools_sitequality.lqual", delim = "\t", col_names = c("chr", "pos", "qual"), skip = 1)

a <- ggplot(var_qual, aes(qual)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()
ggsave("elephants_subset_vcftools_sitequality.tiff", width=15,height=11)
```
A minimum quality of 30 is generally good for filtering for  

>[!IMPORTANT]
> What do we see in our graph? Why might it be low or high? 

### Look at the depth per position 

Plotting mean depth per site
```
R
library(tidyverse)
library (ggplot2)
var_depth <- read_delim("./elephants.chr.11.12.subset_vcftools_depth.lqual", delim = "\t", col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)
a <- ggplot(var_depth, aes(mean_depth)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

ggsave("elephants_subset_vcftools_depth_per_site.tiff", width=15,height=11)
```
This provides some information on where to put a cutoff to remove paralogs. 

### Look at the missingness per site 
```
R
library(tidyverse)
library (ggplot2)
var_miss <- read_delim("./elephants.chr.11.12.subset_vcftools_missingsite.lmiss", delim = "\t", col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), skip = 1)

a <- ggplot(var_miss, aes(fmiss)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
a + theme_light()

ggsave("elephants_subset_vcftools_missingness_per_site.tiff", width=15,height=11)
```
Then we can do a summary within R
```
summary(var_miss$fmiss)
```
> [!IMPORTANT]
> Do we have a lot of missing data at each site? 


### Allele frequencies 
Let's look at the allele frequencies, which show how commona  specific variant of an allele is within a population. You can think of this as the proportion of individuasl that have a specific variant in the population. 
```
R
library(tidyverse)
library (ggplot2)

var_freq <- read_delim("./elephants.chr.11.12.subset_vcftools_allelefreq.frq", delim = "\t", col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1)

# find minor allele frequency
var_freq$maf <- var_freq %>% select(a1, a2) %>% apply(1, function(z) min(z))

#plot
b <- ggplot(var_freq, aes(maf)) + geom_density(fill = "dodgerblue1", colour = "black", alpha = 0.3)
b + theme_light()

ggsave("elephants_subset_vcftools_allele_frequency2.tiff", width=15,height=11)

```

> [!IMPORTANT]
> What do we see? 




