# Read Mapping and Variant Calling 

## Part one: Read Mapping

### File types for Part 1

- fastq files -> this file stores raw sequence data from the sequencing machine. For paired-read sequencing, wew ill obtain a Read 1 (R1) file and Read 1 (R2) file.  

Let's look at our fastq file: 

```
less /anvil/scratch/x-hfigueiro/NA5206_AGTTGGCT-GCAACCAT_L003_n100_R2_001.fastq.gz
```

The four lines correspond to: 
```
1. A sequence identifier with information about the sequencing run and the cluster. 

2. The sequence (the base calls; A, C, T, G and N).

3. A separator, which is simply a plus (+) sign.

4. The base call quality scores. These are Phred +33 encoded, using ASCII characters to represent the numerical quality scores.
```

- Let's look at our BAM file, which stands for Binary Alignment Map:
```
module load samtools 

samtools view /anvil/scratch/x-hfigueiro/NA5206.loxAfr4_NC000934_subsampled.bam | less -S
```

### Setting up our directories

First, let's make a new directory in our home folder: 

``` 
cd #cd to your home directory
mkdir Alignment_SNPcalling
```
And in our scratch folder: 
```
mkdir /anvil/scratch/x-YOURNAME/Results_Alignment_SNPcalling
```
###  Alignment of reads to reference genome

First, we need to align our trimmed fastq files to the reference genome. 

What is a reference genome and where did we get it? We can search for previously published reference genomes on the National Center for Biotechnology Information (NCBI)'s website (https://www.ncbi.nlm.nih.gov/search/) 

- African savannah elephant reference genome: https://www.ncbi.nlm.nih.gov/datasets/taxonomy/9785/



First, we need to make a job file that contains the script we want to run. 

```
cd /home/YOURNAME/Alignment_SNPcalling
touch 01_alignment.sh
```
Then we will copy and paste this information below into our alignment script: 

```
vim 01_alignment.sh
#paste the contents below
wq
```

Script for alignment
```
#!/bin/sh -l
#SBATCH -A bio240351  
#SBATCH -p wholenode 
#SBATCH --nodes=1
#SBATCH --ntasks=1 
#SBATCH --time=1:30:00
#SBATCH --job-name alignment
#SBATCH -o 01_alignment.out
#SBATCH -e 01_alignment.err

module load bwa
module load samtools

R1=/anvil/scratch/x-hfigueiro/NA5206_AGTTGGCT-GCAACCAT_L003_n100_R1_001.fastq.gz
R2=/anvil/scratch/x-hfigueiro/NA5206_AGTTGGCT-GCAACCAT_L003_n100_R2_001.fastq.gz
REF=/anvil/scratch/x-hfigueiro/Chromosomes.v2.fasta
OUTDIR=/anvil/scratch/x-YOURNAME/Results_Alignment_SNPcalling

bwa index ${REF} #index the reference genome of the Elephant

bwa mem -R "@RG\tID:NA5206\tSM:NA5206" ${REF} ${R1} ${R2} | \
samtools view -q 10 -h -b - | \
samtools sort -o ${OUTDIR}/NamibiaElephant_NA5206.bam -
```
Then we can run the script: 
```
sbatch 01_alignment.sh
```


We can view the out file to see how the job is running: 

```
less slurm-9062900.out
```
We can see the job is running. This will take a long time to align the fastq file to the reference genome, so we will use our previously made reference genome. 


### Depth of our bam file

We will calculate the depth of our bam file. The **depth** is the average number of reads that align to a specific position on the reference genome. **Coverage** refers to the percentage of the genome that is coveraged by at least one reads. 

First, we need to create a job file to place our script into: 
```
cd /home/YOURNAME/Alignment_SNPcalling
touch 02_depth.sh
```

Then we will copy and paste this information below into our alignment script: 

```
vim 02_depth.sh
#paste the contents below
wq
```
Script for calculating depth for a bam file: 
```
#!/bin/sh -l
#SBATCH -A bio240351  
#SBATCH -p wholenode 
#SBATCH --nodes=1
#SBATCH --ntasks=1 
#SBATCH --time=1:30:00
#SBATCH --job-name depth
#SBATCH -o 02_depth.out
#SBATCH -e 02_depth.err

module load samtools

BAM=/anvil/scratch/x-hfigueiro/NA5206.loxAfr4_NC000934_subsampled.bam

samtools depth ${BAM}  |  awk '{sum+=$3} END { print "Average = ",sum/NR}' > depth_NA5206.txt
```
Then we can run the script: 
```
sbatch  02_depth.sh
```
Now we can see a new file in our folder called `depth_NA5206.txt`
```
less /anvil/scratch/x-hfigueiro/depth_NA5206.txt
```
What is the depth of our bam file? 



## Part two: Variant Calling with GATK


The GATK (Genome Analysis Toolkit) is one of the most used programs for genotype calling in sequencing data in model and non model organisms. However, the GATK was designed to analyze human genetic data and all its pipelines are optimized for this purpose (https://gatk.broadinstitute.org/hc/en-us).

### Quick background on file types: 


- A fastq.gz file contains the raw reads that are not aligned to the reference genome. Let's see what a fastq.gz file looks like: 
```
less /anvil/scratch/x-hfigueiro/NA5206_AGTTGGCT-GCAACCAT_L003_n100_R2_001.fastq.gz

```
- A bam file is called a Binary Alignment Map and it is a file where the reads of an individual are aligned to a reference genome. So we know where each read goes on the genome. 
```
samtools view /anvil/scratch/x-hfigueiro/NA5206.loxAfr4_NC000934_subsampled.bam | less -S

```

- A variant call format (VCF) file contains the variants along the genome for a single individual, or multiple individuals. 
```
zless -S /anvil/scratch/x-hfigueiro/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1.vcf.gz
```

### HaplotypeCaller
HaplotypeCaller calls SNPs and indels via local de-novo assembly of haplotypes. This means when it encounters a region showing signs of variation, it discards mapping information and reassembles the reads in that region. This just allows HaplotypeCaller to be more accurate wen calling regions that are typically difficult to call. 

First, we will make a new script for Haplotypecaller 
```
touch 03_GATK_HaplotypeCaller.sh
```

Then we can run the script:

```
#!/bin/sh -l
#SBATCH -A bio240351  
#SBATCH -p wholenode 
#SBATCH --nodes=1
#SBATCH --ntasks=1 
#SBATCH --time=1:30:00
#SBATCH --job-name GATK_HaplotypeCaller
#SBATCH -o 03_GATK_HaplotypeCaller.out
#SBATCH -e 03_GATK_HaplotypeCaller.err

module load gatk/4.1.8.1

BAM=/anvil/scratch/x-hfigueiro/NA5206.loxAfr4_NC000934_subsampled.bam

REF=/anvil/scratch/x-hfigueiro/Chromosomes.v2.fasta

gatk HaplotypeCaller \
-I ${BAM} \
-R ${REF} \
-ERC GVCF \
-O /anvil/scratch/x-YOURNAME/Results_Alignment_SNPcalling/NA5206.loxAfr4_NC000934_subsampled
```

Let's sbatch the script and look at our output file to see the job running: 
```
sbatch 03_GATK_HaplotypeCaller.sh
less 03_GATK_HaplotypeCaller.out
```

We can see the job is running, and is 







```
#!/bin/sh -l
#SBATCH -A bio240351  
#SBATCH -p wholenode 
#SBATCH --nodes=1
#SBATCH --ntasks=1 
#SBATCH --time=1:30:00
#SBATCH --job-name GATK_HaplotypeCaller
#SBATCH -o 03_GATK_HaplotypeCaller_CH0893_Botswana.out
#SBATCH -e 03_GATK_HaplotypeCaller_CH0893_Botswana.err

module load gatk/4.1.8.1

BAM=/anvil/scratch/x-hfigueiro/CH0893_Botswana.loxAfr4_NC000934.bam

module load samtools
samtools index /anvil/scratch/x-hfigueiro/CH0893_Botswana.loxAfr4_NC000934_subsample.bam

REF=/anvil/scratch/x-hfigueiro/Chromosomes.v2.fasta

gatk HaplotypeCaller \
-I ${BAM} \
-R ${REF} \
-ERC GVCF \
-O /anvil/scratch/x-YOURNAME/Results_Alignment_SNPcalling/CH0893_Botswana.loxAfr4_NC000934_subsampled.gvcf
```
### GenomeDBImport
This step creates a GenomicsDB datastore to speedup joint genotyping. This in prep for the Genotype calling.

First, we will make a script: 
```
touch 04_GATK_GenomeDBImport.sh
```

There's a few things we need to do in order to prepare for GenomeImport. We need to make a sample map: 
```
less samplemap.txt

SE2109_Tanzania /anvil/scratch/x-hfigueiro/SE2109_Tanzania.loxAfr4_NC000934_subsampled.gvcf
CH0893_Botswana /anvil/scratch/x-hfigueiro/CH0893_Botswana.loxAfr4_NC000934_subsampled.gvcf
NA5206_Namibia  /anvil/scratch/x-hfigueiro/NA5206.loxAfr4_NC000934_subsampled.gvcf
```

Then we will paste this script below: 
```
#!/bin/sh -l
#SBATCH -A bio240351  
#SBATCH -p wholenode 
#SBATCH --nodes=1
#SBATCH --ntasks=1 
#SBATCH --time=1:30:00
#SBATCH --job-name GATK_GenomeDBImport
#SBATCH -o 04_GATK_GenomeDBImport.out
#SBATCH -e 04_GATK_GenomeDBImport.err

module load gatk/4.1.8.1

OUTDIR=/anvil/scratch/x-lhennelly/ConGen2025/GenomeImport
TEMPDIR=/anvil/scratch/x-lhennelly/ConGen2025/GenomeImport_Temp
SAMPLEMAP=/home/x-lhennelly/Day1/samplemap.txt

gatk GenomicsDBImport --genomicsdb-workspace-path ${OUTDIR}/elephant_gvcf_db --batch-size 50 -L chr1 --sample-name-map ${SAMPLEMAP} --tmp-dir ${TEMPDIR}
```

And we can run the file: 
```
sbatch 04_GATK_GenomeDBImport.sh
```

Let's see how its working. We can cd into the GenomeDBImport:
```
cd /anvil/scratch/x-lhennelly/ConGen2025/GenomeImport
ls
```
We should see the GenomeBPImport start making the databases by chromosome

### GenotypeGVCFs
Now we can perform joint genotyping that were pre-called with HaplotypeCaller.

- It will look at the available information for each site from both variant and non-variant alleles across all samples, and then produce a multi-sample VCF that contains only sites that are found as variant in at least one sample.

https://gatk.broadinstitute.org/hc/en-us/articles/360037057852-GenotypeGVCFs

Let's make a directory for our VCF output:
```
mkdir /anvil/scratch/x-lhennelly/ConGen2025/GenotypeGVCF
mkdir /anvil/scratch/x-lhennelly/ConGen2025/GenotypeGVCF_scratch
```
Now we can make a job file and paste our script into it: 
```
touch 05_GATK_GenotypeGVCFs.sh
```

Now let's run the job file:
```
#!/bin/sh -l
#SBATCH -A bio240351  
#SBATCH -p wholenode 
#SBATCH --nodes=1
#SBATCH --ntasks=1 
#SBATCH --time=13:30:00
#SBATCH --job-name GenotypeGVCF
#SBATCH -o 05_GenotypeGVCF.out
#SBATCH -e 05_GenotypeGVCF.err

module load gatk/4.1.8.1

REF=/anvil/scratch/x-hfigueiro/Chromosomes.v2.fasta
OUTDIR=/anvil/scratch/x-lhennelly/ConGen2025/GenotypeGVCF
TEMPDIR=/anvil/scratch/x-lhennelly/ConGen2025/GenotypeGVCF_scratch

cd /anvil/scratch/x-lhennelly/ConGen2025/GenomeImport

gatk GenotypeGVCFs -R ${REF} -V gendb://elephant_gvcf_db -O ${OUTDIR}/gatk_Elephants.vcf.gz --tmp-dir ${TEMPDIR}
```
And we can look at our results: 







