# PCA and Admixture 

We will use Principle Component Analysis (PCA) and estimate individual admixture proportions with Admixture to investigate population structure. 

## Principle Component Analysis

First, we will investigate population structure using principal component analysis. 

- PCA aims to identify the main axes of variation in a dataset with each axis being independent of the next. The first component summarizes the major axis variation and the second the next largest, and so on, until all the cummulative variation is explained. 
- For genomic data, PCA summarizes the major axes of variation in allele frequencies

To perform PCA, we will use plink (https://www.cog-genomics.org/plink/), a useful program to analyze population genomic data. 

## Setting us up for analyses

First, lets cd into our directory for our analyses. 
```
cd /anvil/scratch/x-YOURNAME/04_population_structure
```

###  Our dataset 
We will be using a VCF of 199 elephant individuals, which include African Savannah Elephants and African Forest Elephants. This amount of indivdiuals will allow us to investigate the population structure of elephants in Africa at a fine-scale. 

Let's first take a look at our VCF file: 
```
zless -S /anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1.vcf.gz
```

### Filtering our raw VCF file 

#### Check for missingness per individual

First, we need to check out raw VCF file and filter our raw VCF file to contain high quality SNPs. That is, SNPs that we are confident that are correct. 

To do this, we will use VCFtools, which is a very common program used to filter variants and samples in our dataset. 

First,  we will check to see the missingness per individual in our raw VCF file. 

To do this, we will work with the `01_checkmissingness.sh` file. 

Create the script: 
```
touch 01_checkmissingness.sh
```
And here's the script: 
```
#!/bin/bash
#SBATCH --job-name 01_checkmissingness
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/01_checkmissingness.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/01_checkmissingness.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load vcftools

VCF=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1.vcf.gz

vcftools --gzvcf ${VCF} --missing-indv
```
Now we can sbatch the script and check the output file in a few minutes: 
```
sbatch 01_checkmissingness.sh
ls
less out.imiss
less /anvil/projects/x-bio240351/shared_data/outputs/out.imiss
```
Here's the output: 
```
INDV    N_DATA  N_GENOTYPES_FILTERED    N_MISS  F_MISS
AM0001  2533335 0       915     0.000361184
AM0003  2533335 0       30890   0.0121934
AM0006  2533335 0       530     0.00020921
AM0008  2533335 0       1931    0.000762236
AM0037  2533335 0       1345    0.000530921
AM4558  2533335 0       484     0.000191053
AM4559  2533335 0       4450    0.00175658
AM4583  2533335 0       720     0.00028421
BA0001  2533335 0       660     0.000260526
BA0003  2533335 0       582     0.000229737
BA0011  2533335 0       562     0.000221842
BE4036  2533335 0       1017    0.000401447
BE4037  2533335 0       1377    0.000543552
BE4039  2533335 0       5554    0.00219237
BE4059  2533335 0       517     0.000204079
CH0888  2533335 0       590     0.000232895
CH0892  2533335 0       1312    0.000517894
CH0893  2533335 0       15254   0.00602131
CH0895  2533335 0       805     0.000317763
CH0935  2533335 0       732     0.000288947
CH0954  2533335 0       393     0.000155131
CH0955  2533335 0       1882    0.000742894
CH0962  2533335 0       1224    0.000483158
DS1506  2533335 0       11604   0.00458052
DS1507  2533335 0       13788   0.00544263
DS1508  2533335 0       10997   0.00434092
DS1509  2533335 0       13048   0.00515052
DS1511  2533335 0       13126   0.00518131
DS1512  2533335 0       9501    0.00375039
DS1514  2533335 0       9490    0.00374605
DS1517  2533335 0       13155   0.00519276
DS1520  2533335 0       10950   0.00432237
DS1531  2533335 0       27027   0.0106685
DS1543  2533335 0       10958   0.00432552
DS1546  2533335 0       14086   0.00556026
DS1548  2533335 0       10730   0.00423552
DS1555  2533335 0       12138   0.00479131
DS1557  2533335 0       11329   0.00447197
ER0022  2533335 0       429     0.000169342
GO0045  2533335 0       5788    0.00228474
GO00T2  2533335 0       4583    0.00180908
GO0G13  2533335 0       1484313 0.585913
GO0G14  2533335 0       1812428 0.715432
GO0G15  2533335 0       1735570 0.685093
GO0G16  2533335 0       2009931 0.793393
GO0G17  2533335 0       3819    0.0015075
GO0G18  2533335 0       6453    0.00254724
GO0G19  2533335 0       3663    0.00144592
GO0G20  2533335 0       4110    0.00162237
GO0G21  2533335 0       3499    0.00138118
GO0G22  2533335 0       3923    0.00154855
GO2981  2533335 0       3853    0.00152092
GO2982  2533335 0       3766    0.00148658
GO2983  2533335 0       12510   0.00493815
GO2984  2533335 0       2621    0.0010346
GO2985  2533335 0       3615    0.00142697
GR0013  2533335 0       8945    0.00353092
GR0018  2533335 0       8014    0.00316342
GR0020  2533335 0       179937  0.0710277
GR0021  2533335 0       1285    0.000507237
GR0022  2533335 0       12250   0.00483552
GR0023  2533335 0       17559   0.00693118
GR0025  2533335 0       18730   0.00739342
GR0026  2533335 0       14685   0.00579671
GR0035  2533335 0       579480  0.228742
GR0036  2533335 0       11549   0.00455881
GR0037  2533335 0       340186  0.134284
GR0038  2533335 0       11729   0.00462987
GR0039  2533335 0       11567   0.00456592
GR0040  2533335 0       11299   0.00446013
GR0041  2533335 0       9569    0.00377723
GR0042  2533335 0       12625   0.00498355
HW0059  2533335 0       776     0.000306316
HW0087  2533335 0       3804    0.00150158
HW0151  2533335 0       633     0.000249868
KE4516  2533335 0       10483   0.00413802
KE4517  2533335 0       1166    0.000460263
KE4540  2533335 0       1228    0.000484737
KE4546  2533335 0       1420    0.000560526
KE4612  2533335 0       5730    0.00226184
KEHIDA  2533335 0       8400    0.00331579
KEWATO  2533335 0       37774   0.0149108
KR0005  2533335 0       516     0.000203684
KR0006  2533335 0       523     0.000206447
KR0007  2533335 0       754     0.000297631
KR0008  2533335 0       1433    0.000565658
KR0011  2533335 0       1307    0.000515921
KR0014  2533335 0       1402    0.000553421
KR0020  2533335 0       1366    0.00053921
KR0027  2533335 0       421     0.000166184
KR0068  2533335 0       735     0.000290131
KR0116  2533335 0       1422    0.000561315
KR0138  2533335 0       406     0.000160263
KRSWAZ  2533335 0       29266   0.0115524
KV4688  2533335 0       372530  0.147051
KV4691  2533335 0       410139  0.161897
KV4697  2533335 0       661839  0.261252
KV4704  2533335 0       209425  0.0826677
LO3503  2533335 0       11456   0.0045221
LO3505  2533335 0       1502994 0.593287
LO3507  2533335 0       18624   0.00735157
LO3508  2533335 0       29707   0.0117264
LO3509  2533335 0       12165   0.00480197
LO3511  2533335 0       11930   0.00470921
LO3513  2533335 0       13047   0.00515013
LO3515  2533335 0       14700   0.00580263
LO3517  2533335 0       12612   0.00497842
LO3522  2533335 0       1464771 0.578199
MA0804  2533335 0       3680    0.00145263
MA0806  2533335 0       551     0.0002175
MA0807  2533335 0       485     0.000191447
MA0811  2533335 0       397     0.00015671
MA0813  2533335 0       923     0.000364342
MA0819  2533335 0       549     0.00021671
MA0821  2533335 0       454     0.00017921
MA0822  2533335 0       1263    0.000498552
MA0824  2533335 0       1100    0.00043421
MF2814  2533335 0       615160  0.242826
MF2816  2533335 0       899919  0.355231
MF2817  2533335 0       992193  0.391655
MF2818  2533335 0       480529  0.189682
MI0002  2533335 0       3872    0.00152842
MI0003  2533335 0       3759    0.00148381
NA4651  2533335 0       911     0.000359605
NA4658  2533335 0       1340    0.000528947
NA4674  2533335 0       381     0.000150395
NA4685  2533335 0       909     0.000358816
NA4702  2533335 0       537081  0.212006
NA4704  2533335 0       1534    0.000605526
NA4708  2533335 0       758     0.00029921
NA5205  2533335 0       1243195 0.490735
NA5206  2533335 0       888847  0.35086
NG2178  2533335 0       697     0.000275131
NG2179  2533335 0       1315    0.000519079
NG2180  2533335 0       604     0.000238421
OD0001  2533335 0       10688   0.00421894
QE0870  2533335 0       1239276 0.489188
QE0879  2533335 0       461425  0.182141
QE0881  2533335 0       681478  0.269004
QE0882  2533335 0       916078  0.361609
QE0883  2533335 0       422227  0.166668
QE0887  2533335 0       1175262 0.463919
QE0888  2533335 0       358612  0.141557
QE2796  2533335 0       1214886 0.47956
QE2798  2533335 0       765792  0.302286
QE2799  2533335 0       449640  0.177489
QE2800  2533335 0       1059680 0.418294
QE2801  2533335 0       339995  0.134208
QE4714  2533335 0       581395  0.229498
QE4717  2533335 0       1068930 0.421946
QE4719  2533335 0       801461  0.316366
SA0976  2533335 0       471     0.000185921
SA0994  2533335 0       178310  0.0703855
SA0995  2533335 0       2238    0.00088342
SE2051  2533335 0       1085    0.000428289
SE2098  2533335 0       3826    0.00151026
SE2101  2533335 0       1329    0.000524605
SE2103  2533335 0       741     0.0002925
SE2109  2533335 0       9861    0.0038925
SE2157  2533335 0       640     0.000252631
SE2162  2533335 0       759     0.000299605
SE2164  2533335 0       570     0.000225
SL0001  2533335 0       60298   0.0238018
SW0901  2533335 0       311     0.000122763
SW0902  2533335 0       438     0.000172895
SW0903  2533335 0       4792    0.00189158
TA1138  2533335 0       860     0.000339473
TA1143  2533335 0       1399    0.000552236
TA1429  2533335 0       639     0.000252237
TA1435  2533335 0       558     0.000220263
TA1439  2533335 0       481     0.000189868
TA1441  2533335 0       749     0.000295658
TA1444  2533335 0       1120    0.000442105
TA1450  2533335 0       602     0.000237631
TA1452  2533335 0       1198    0.000472894
TA1454  2533335 0       556     0.000219474
TA1463  2533335 0       1200    0.000473684
TA1467  2533335 0       561     0.000221447
WA4004  2533335 0       1024    0.00040421
WA4006  2533335 0       647     0.000255395
WA4010  2533335 0       2020    0.000797368
WA4013  2533335 0       893     0.0003525
WA4018  2533335 0       654     0.000258158
WA4020  2533335 0       1461    0.00057671
WA4021  2533335 0       1008    0.000397894
WA4023  2533335 0       707     0.000279079
WA4024  2533335 0       1024    0.00040421
ZZ0145  2533335 0       1132    0.000446842
ZZ0147  2533335 0       1268    0.000500526
ZZ0148  2533335 0       673     0.000265658
ZZ0149  2533335 0       1203    0.000474868
ZZ0150  2533335 0       439     0.000173289
ZZ0153  2533335 0       352     0.000138947
ZZ0156  2533335 0       5536    0.00218526
ZZ0157  2533335 0       459     0.000181184
ZZ0160  2533335 0       1275    0.000503289
ZZ0161  2533335 0       25282   0.00997973
ZZ0162  2533335 0       1197    0.0004725
ZZCHRI  2533335 0       12787   0.0050475
```
We can see some of the elephant individuals are missing quite alot of data! Such as NA5205 is missing around 49% of called sites. We will remove elephant individuals that are missing more than 20% of data in our next script. 

#### Keeping high quality SNPs

Next, we will filter the raw VCF to keep the best quality SNPs and remove indels. We will use the script below with VCFtools.

Create the script: 
```
touch 02_filterSNPs.sh
```

We will then use grep to make a list of elephants that are missing over 20% of their data: 
```
awk -F'\t' '$5 > 0.20' out.imiss > Elephant_to_remove.txt
awk '{print $1}' Elephant_to_remove.txt > Elephant_to_remove_names.txt
sed '1d' Elephant_to_remove_names.txt > Elephant_to_remove_names_final.txt
```
And here's the script to remove the Elephants: 
```
#!/bin/bash
#SBATCH --job-name 02_filterSNPs
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/02_filterSNPs.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/02_filterSNPs.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load vcftools

VCF=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1.vcf.gz
OUT=/anvil/scratch/x-YOURNAME/04_population_structure/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8

vcftools --gzvcf ${VCF} --minQ 30 --remove-indels --max-missing 0.8 --remove /anvil/scratch/x-hfigueiro/Elephants_to_remove.txt --recode --recode-INFO-all --out ${OUT}
```
- remove-indels -> remove small indels (insertions and deletions)
- max-missing 0.8 -> remove SNPs were they are missing 80% of individuals at that site
- minQ30 -> keep SNps with a minimum quality of 30

And we can run the command: 
```
sbatch 02_filterSNPs.sh
```
Meaning of the flags: 
- minQ 30 -> keep sites that have at least a minimum quality score of 30. This means we think think the base quality accuracy at this position is 99.9% correct.

- remove-indels -> this removes indels, which are short insertions and deletions. For hte rest of our analyses, we are just interested in SNPs.

- max-missing 0.8 -> keep positions where there is data for at least 80% of individuals
  
- remove Elephants_to_remove.txt -> this provides a list of sample ID's to remove from the VCF, which were samples that had more than 20% missing data

This step will take a couple of minutes. 
Let's compare the output files from the raw VCF and our filtered VCF: 
```
From our unfiltered dataset, we had 2533335 SNPs, as we can see from the imiss file. 

less /anvil/scratch/YOUR_USERNAME/log/02_filterSNPs/02_filterSNPs.err 

It states "After filtering, kept 175 out of 199 Individuals and After filtering, kept 2505808 out of a possible 2533335 Sites"
```
We can see the filtering worked! 

### Use PLINK to remove linked sites

Plink (https://zzz.bwh.harvard.edu/plink/) is a commonly used program to do conduct various population genomic analyses. Mainly, people use it for doing PCA, make input files for other programs, and quantify relatedness. 

An assumption of PCA is that we use independent data. However, genomic data and allele frequencies can be correlated due to physical linkage. We will remove sites that are within a certain distance to each other to limit linked sites.  

First we will open the plink script: 
```
touch 03_removeLD.sh
```

Then we can use this script: 

```
#!/bin/bash
#SBATCH --job-name 03_removeLD
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/03_removeLD.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/03_removeLD.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

VCF=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_IDs.recode.vcf.vcf
OUT=/anvil/scratch/x-YOURNAME/04_population_structure/LoxAfr4_elephant_filtered_LD
FINAL=/anvil/scratch/x-YOURNAME/04_population_structure/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD

module load biocontainers/default
module load plink/1.90b6.21

plink --vcf ${VCF} --indep-pairwise 50 5 0.5 --out ${OUT} --const-fid 0 --allow-extra-chr
plink --vcf ${VCF}  --extract ${OUT}.prune.in  --out ${FINAL} --recode --const-fid 0 --allow-extra-chr

```
And now we can run the script and look at the output files: 
```
sbatch 03_removeLD.sh
less 03_removeLD.out
```

Let's look at the plink output files to understand the file structure:

```
less /anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD.map
less /anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD.ped
```


### PCA!!!!! 

Woo! Now we can infer our PCA. We will use plink to infer the PCA.

First we will make a new script to run the PCA command: 
```
touch 04_runPCA.sh
```

Then we will copy and paste this script below: 
```
#!/bin/bash
#SBATCH --job-name 04_runPCA
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/04_runPCA.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/04_runPCA.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load biocontainers/default
module load plink/1.90b6.21

FILE=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD

plink --file ${FILE} --pca var-wts  --const-fid 0 --allow-extra-chr --out PCA_Elephants
```
Then we will run the script and take a look at the output files: 
```
sbatch 04_runPCA.sh
less PCA_Elephants.eigenval #tells us the contribution of each PC in explaining the variation
less PCA_Elephants.eigenvec #Principal components
```

We will now open this in R to plot. Put the file "PCA_Elephants.csv" into your R directory.

### Plotting the results 

First, we will plot a map of our samples and where their origins are from using R. In general, R is a powerful way to plot and analyze geospatial data! 

```
cp /anvil/projects/x-bio240351/shared_data/outputs/PCA_Elephants.csv /anvil/scratch/x-YOURNAME/04_population_structure/
```
Let's open R on your own computer
```
R
library(ggplot2)
install.packages("maps", lib="/anvil/projects/x-bio240351/shared_data/R")
library(maps, lib.loc="/anvil/projects/x-bio240351/shared_data/R")
install.packages("mapdata", lib="/anvil/projects/x-bio240351/shared_data/R")
library(mapdata, lib.loc="/anvil/projects/x-bio240351/shared_data/R")
install.packages("rworldmap", lib="/anvil/projects/x-bio240351/shared_data/R")
library(rworldmap, lib.loc="/anvil/projects/x-bio240351/shared_data/R")

# First we will plot the data:

#Plot world map
world <- map_data('world')
p <- ggplot(world, aes(lat, long)) +
    geom_map(map=world, aes(map_id=region), fill="gray95", color="darkgray") +
    coord_quickmap() + theme_classic()  +
    coord_sf(xlim = c(-30, 60 ), ylim = c(-50, 50), expand = TRUE) 

# Add coordinates
q <- p +  geom_point(data = dat, 
             aes(x = Long, y = Lat, fill=Species), 
             alpha = 1, 
             size = 3.5, shape = 21) + scale_fill_manual(values = c("mediumseagreen","orange2"))
q
```

Now we will plot the PCA results
```
#in R
install_packages("ggplot2")
library(ggplot2)
install.packages("ggrepel")
library(ggrepel)

#Read in data 
dat <- read.csv("PCA_Elephants.csv")

#plot graph of Forest vs Savannah Elephants                                      
p <- ggplot(dat,aes(x=PC1,y=PC2,color=Species)) + geom_point(size = 3) + theme_classic() + scale_color_manual(values = c("mediumseagreen","orange2"))  
p

#plot graph by country  
p <- ggplot(dat,aes(x=PC1,y=PC2,color=Country)) + geom_point(size = 3) + theme_classic() + scale_color_manual(values = c("mediumseagreen","orange2"))  
p

#plot Eigenvalues 

Eigenvalues <- c(0.247388568, 0.057569039, 0.052510705, 0.040342562, 0.039821937, 0.039464296, 0.039378169)
barplot(Eigenvalues)

```
What biological or ecological fact we know about African elephants explains most of the variation in the genomic dataset? 

## Individual Admixture Proportions using Admixture 

Now we can infer individual admixture proportions with Admixture. This is a genetic clustering program to define populations and assign individuals to them.

We will use the plink files we just make to run Admixture.

Let's make a job file: 
```
cd /anvil/scratch/x-YOURNAME/04_population_structure
touch 05_Admixture.sh
```
And we can paste this script into the job file: 
```
#!/bin/bash
#SBATCH --job-name 05_Admixture
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/05_Admixture.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/05_Admixture.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load biocontainers/default
module load admixture 
module load plink

INPUT=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD
OUTPUT=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD_bed

plink --file ${INPUT} --make-bed --allow-extra-chr --const-fid 0 --out ${OUTPUT}

FILE=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD_bed.bed

for i in {2..5}
do
admixture --cv ${FILE} $i > log${i}.out
done
```
We will take a look at our output files: 
```
less log2.out
less /anvil/projects/x-bio240351/shared_data/outputs/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_filtered_LD_bed.2.Q
```

### Plotting the results in R

We will now move onto plotting our results in R Studio. Make sure you have the CSV files in your R directory, and we will load them into our R directory for the class. 

First we will plot the individual ancestry proportions at K=2 to K=4 for the African Elephants. 
```
R

# Install and load packages
install_packages("tidyverse")
library(tidyverse)
install_packages("dplyr")
library(dplyr)
install_packages("ggplot2")
library (ggplot2)

###################
# Results at K=2 ##
###################
dat <-read.csv ("Admixture_K2.csv", header=TRUE)

## Organize the dataset
data_long <- gather(dat, Admixture, Percentage, Ancestry1:Ancestry2, factor_key=TRUE)

# Plot admixture at K=2
q <- ggplot(data_long, aes(fill=Admixture, y=Percentage, x=Sample)) + 
    geom_bar(position="stack", stat="identity") +scale_fill_manual(values = c("mediumseagreen","orange2", "orange2", "royalblue4", "red", "honeydew4"))  + theme(axis.text.x = element_text(angle = 90, hjust = 1))
q

###################
# Results at K=3 ##
###################
dat <-read.csv ("Admixture_K3.csv", header=TRUE)

## Organize the dataset
data_long <- gather(dat, Admixture, Percentage, Ancestry1:Ancestry3, factor_key=TRUE)

# Plot admixture at K=3
q <- ggplot(data_long, aes(fill=Admixture, y=Percentage, x=Sample)) + 
    geom_bar(position="stack", stat="identity") +scale_fill_manual(values = c("orange2","mediumseagreen", "yellow2"))  + theme(axis.text.x = element_text(angle = 90, hjust = 1))
q

###################
# Results at K=4 ##
###################
dat <-read.csv ("Admixture_K4.csv", header=TRUE)

## Organize the dataset
data_long <- gather(dat, Admixture, Percentage, Ancestry1:Ancestry4, factor_key=TRUE)

# Plot admixture at K=4
q <- ggplot(data_long, aes(fill=Admixture, y=Percentage, x=Sample)) + 
    geom_bar(position="stack", stat="identity") +scale_fill_manual(values = c("orange2","mediumseagreen", "greenyellow", "darkgreen"))  + theme(axis.text.x = element_text(angle = 90, hjust = 1))
q

```

Next, we will plot the sample map with each location of an individual elephant showing its individual admixture porportions. Adding a geographical component helps us understand how Elephant population structure is structured across the landscape. 

module load biocontainers/default
 module load r/4.1.0 
```
# Install and load packages
install_packages("rnaturalearth")
library("rnaturalearth")
install_packages("rnaturalearthdata")
library("rnaturalearthdata")
install_packages("maps")
library(maps)
install_packages("ggplot2")
library(ggplot2)
install.packages("scatterpie")
library(scatterpie)

install.packages("scatterpie", lib="/home/x-lhennelly/Rpackages")
)


/home/x-lhennelly/Rpackages

/tmp/RtmpNH8VMo/downloaded_packages


########################
#### Plot world map ####
########################
world <- map_data('world')
p <- ggplot(world, aes(lat, long)) +
    geom_map(map=world, aes(map_id=region), fill="gray95", color="darkgray") +
    coord_quickmap() + theme_classic()  +
    coord_sf(xlim = c(-30, 60 ), ylim = c(-50, 50), expand = TRUE) 
 

################
### Plot K=2 ###
################
dat <- read.csv ("Admixture_K2_map.csv", header=TRUE) #read dataset

#Plot the elephant locations and their data onto the map
p + geom_scatterpie(aes(x=long, y=lat, r=1),
                    data=dat, cols=c("A", "B"), color="black") + scale_fill_manual(values = c("mediumseagreen","orange2"))

################
### Plot K=3 ###
################
dat <- read.csv ("Admixture_K3_map.csv", header=TRUE) #read dataset

 #Plot the elephant locations and their data onto the map
p + geom_scatterpie(aes(x=long, y=lat, r=1),
                    data=dat, cols=c("A", "B", "C"), color="black") + scale_fill_manual(values = c("orange2","mediumseagreen", "yellow"))

################
### Plot K=4 ###
################
dat <- read.csv ("Admixture_K4_map.csv", header=TRUE) #read dataset

#Plot the elephant locations and their data onto the map
p + geom_scatterpie(aes(x=long, y=lat, r=1),
                    data=dat, cols=c("A", "B", "C", "D"), color="black") + scale_fill_manual(values = c("orange2","yellow", "mediumseagreen", "lightgreen"))

################
### Plot K=5 ###
################
dat <- read.csv ("Admixture_K5_map.csv", header=TRUE) #read dataset

 #Plot the elephant locations and their data onto the map
p + geom_scatterpie(aes(x=long, y=lat, r=1),
                    data=dat, cols=c("A", "B", "C", "D", "E"), color="black") + scale_fill_manual(values = c("mediumseagreen","lightgreen", "yellow", "orange2", "darkgreen"))
```
How does geography or ecology relate to our results using Admixture?  


# Fst across the genome of Elephants 

We will use Fst across the genome to estimate 

```
touch 06_Fstscan.sh
```
Paste below into script:
```
#!/bin/bash
#SBATCH --job-name 06_Fstscan
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=1:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/x-lhennelly/log/06_Fstscan.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/x-lhennelly/log/06_Fstscan.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

module load vcftools

VCF=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_minQ30_noindels_maxmissing0.8_IDs.recode.vcf.vcf
forestelephant=/anvil/projects/x-bio240351/shared_data/vcf/forestelephant_list.txt
savannahelephant=/anvil/projects/x-bio240351/shared_data/vcf/savannahelephant_list.txt
OUT=/anvil/scratch/x-YOURNAME/shared_data/04_population_structure/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_ForestvsSavannah


## Between forest and savannah elephants
vcftools --vcf ${VCF} \
--weir-fst-pop ${forestelephant} \
--weir-fst-pop ${savannahelephant} \
--fst-window-size 100000 \
--fst-window-step 10000 \
--out ${OUT}

###############################
## Within savannah elephants ##
###############################
savannahelephant_KenyaTanzania=/anvil/projects/x-bio240351/shared_data/vcf/savannahele_KenyaTanz.txt
savannahelephant__Namibia_Botswana=/anvil/projects/x-bio240351/shared_data/vcf/savannahele__Namibia_Botswana.txt
OUT_savannah=/anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_withinSavannah

vcftools --vcf ${VCF} \
--weir-fst-pop ${savannahelephant_KenyaTanzania} \
--weir-fst-pop ${savannahelephant__Namibia_Botswana} \
--fst-window-size 100000 \
--fst-window-step 10000 \
--out ${OUT_savannah}
```

Let's look at some of our results: 

```
less /anvil/projects/x-bio240351/shared_data/vcf/LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_ForestvsSavannah.windowed.weir.fst
```
And remove the -NAN
```
sed '/n/d' LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_ForestvsSavannah.windowed.weir.fst >  LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_ForestvsSavannah.windowed_noNAN.weir.fst

sed '/n/d' LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_withinSavannah.windowed.weir.fst >  LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_withinSavannah.windowed_noNAN.weir.fst


```
Now we can calculate the average Fst value between Forest and Savannah Elephants 
```
awk '{ sum += $6 } END { print "Average Fst between Savannah and Forest Elephants:", sum/NR }' LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_ForestvsSavannah.windowed_noNAN.weir.fst > averageFst_Savannah_Forest_Elephants.txt

awk '{ sum += $6 } END { print "Average Fst between Savannah Kenya and Savannah Namibia Elephants:", sum/NR }' LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_withinSavannah.windowed_noNAN.weir.fst > averageFst_Savannah_to_Savannah_Elephants.txt

head -n 1000 LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_withinSavannah.windowed_noNAN.weir.fst > LoxAfr4_elephant_no1stdegree_variable_sites_nomultiallelics_noindels_10dp_3het_chr1_withinSavannah.windowed_noNAN_reduced.weir.fst

```
What's the average Fst? 

Now we can plot the Fst betwen Savannah and Forest Elephants across the elephant genome 
```
module load r/4.4.1
R
library (ggplot2)

dat <- read.csv ("Fst_Savannah_Forest_elephant.csv", header=TRUE) #read dataset

#plot data
x <- ggplot(dat, aes(x=BIN_START, y=WEIGHTED_FST, color=Comparison)) + theme_classic() + geom_line(size=1.1) 
x


```









