# ConGen 2025 - Genetic diversity tutorial

## Introduction

Heterozygosity is a simple and informative statistic that can be obtained by analyzing whole-genome data. You can calculate the average heterozygosity of an individual or assess the local heterozygosity throughout the genome to answer more specific questions.

There are many methods available to achieve this. In this tutorial, we will use two strategies. To obtain the average heterozygosity, we will manipulate the VCF file using bcftools. To explore heterozygosity variation across the genome, we will use ANGSD and other associated tools.

---------------------------------------------------------------------------

### Section 1 - Average heterozygosity

#### a) Use the following script to calculate the average heterozygosity per individual

The bash script below calculates the heterozygosity of each sample in a VCF file. It uses the bcftools command line tool to extract the list of sample names and calculate the number of heterozygous variants for each sample. The results are written to a tab-separated file with the name "heterozygosity.tsv", which contains two columns: "Sample" and "Heterozygosity". Then, the last command line add a last column with population origin of each individual. Reminders:
- Go to your directory `/anvil/scratch/YOUR_USERNAME/05_genetic_diversity`
- Create a new file with a name that ends with `.sh` using vim/nano (e.g. `01_run_avg_het.sh`),
- Copy the script below to the file,
- Edit it to add your own `YOUR_USERNAME`,
- Run it on Anvil using `sbatch <job_name>.sh`, and
- Check the job status with: `squeue -u <your_username>`.

```bash
#!/bin/bash
#SBATCH --job-name avg_het
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=0:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/avg_het.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/avg_het.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

#Load bcftools
module load biocontainers/default
module load bcftools/1.17

#Define path to input files - do not change it
VCF_FILE="/anvil/projects/x-bio240351/shared_data/vcf/elephants.chr.11.12.subset.vcf.gz"
POPFILE="/anvil/projects/x-bio240351/shared_data/popfile.txt"
GENOME_LENGTH="152468515" #Sum of the chromosome lengths included in this analysis

#Define the output name. Replace it with your actual username
OUTPUT_FILE="/anvil/scratch/YOUR_USERNAME/05_genetic_diversity/elephants_heterozygosity.tsv"

#Get a list of sample names from the VCF file
SAMPLES=$(bcftools query -l $VCF_FILE)

#Write a header line to the output file
echo -e "Sample\tHeterozygous_sites\tHeterozygosity" > $OUTPUT_FILE

#Loop through each sample and calculate the heterozygosity
for SAMPLE in $SAMPLES; do
  HETEROZYGOUS=$(bcftools view -s $SAMPLE $VCF_FILE | grep -v "#" | grep -o "0/1" | wc -l)
  HETEROZYGOSITY=$(echo "scale=7; $HETEROZYGOUS / $GENOME_LENGTH" | bc)
  echo -e "$SAMPLE\t$HETEROZYGOUS\t$HETEROZYGOSITY" >> $OUTPUT_FILE
done

#Add population origin of each sample to the heterozygosity file
awk 'NR==FNR {pop[$1]=substr($0, index($0, $2)); next} FNR==1 {print $0"\tPopulation"} FNR>1 {print $0"\t"pop[$1]}' $POPFILE $OUTPUT_FILE > temp && mv temp $OUTPUT_FILE
```

#### b) Plot the results using the R

Follow the steps below to load R in the Anvil server and plot the results
- load the R module by entering `module load r/4.1.0` in the terminal.
- Then, start an R session by typing `R`.
- Copy and paste the commands in the script below in the terminal. Make sure you understand each command in the R script with the help of the explanations provided in the commented lines (#).
- Use `quit()` to close R.
- Use Anvil browser to visualize the output file `heterozygosity_boxplot_by_population.png`.
  
In summary, this R code block generates a boxplot of heterozygosity values, grouping individuals by population, using data from a tab-separated value (TSV) file. The ggplot2 library is loaded, and the name of the input file is specified. The data is read in using the read.table function, and a boxplot is created using ggplot with "Populations" on the x-axis and "Heterozygosity" on the y-axis. The resulting plot is saved as a png file named "heterozygosity_boxplot_by_population.png".

```r
# Call ggplot library
library(ggplot2)

# Define the input file
INPUT_FILE <- "elephants_heterozygosity.tsv"

# Read in the data from the input file
data <- read.table(INPUT_FILE, header=TRUE, sep="\t")

# Create a boxplot for heterozygosity by population
plot <- ggplot(data, aes(x=Population, y=Heterozygosity, color=Population)) +
  geom_boxplot(outlier.shape = NA, width=0.6, alpha=0.6, color="black") +  # Boxplot without outliers, adjust box width and transparency
  geom_jitter(size=2, width=0.1, alpha=0.7) +  # Jittered dots for each individual with some horizontal spacing
  theme(axis.text.x=element_text(angle=45, hjust=1)) +  # Angle the x-axis labels for better readability
  labs(x="Population", y="Heterozygosity", color="Population") +
  scale_color_discrete(name="Population")  # Color legend for populations

# Save the plot to a file
ggsave("heterozygosity_boxplot_by_population.png", plot, width=10, height=6, dpi=300)
```

> [!IMPORTANT]
> :elephant::grey_question: How does the genetic diversity of savanna elephants compare to that of forest elephants, and what factors might contribute to these differences? What could explain the observed levels of heterozygosity in the savanna elephants of Queen Elizabeth National Park (QENP), Uganda?
  
---------------------------------------------------------------------------

### Section 2 - Genome-wide heterozygosity

Another way to display heterozygosity is by using a window-based approach. This approach allows for the visualization of variation throughout the genome and the ability to focus on regions of particularly high or low diversity. It also permits comparison of results with other methods, such as Runs of Homozygosity (RoH). There are several software programs that can perform this task in a similar way, including ANGSD.

ANGSD (Analysis of Next Generation Sequencing Data) is an open-source software designed to analyze low-depth Next Generation Sequencing (NGS) data. Developed by researchers at the University of Copenhagen, ANGSD was specifically designed to handle large-scale sequencing data, particularly from non-model organisms. ANGSD provides a flexible and efficient framework for analyzing complex genomic data, enabling researchers to perform various types of population genetics and genomics analyses.

Key capabilities of ANGSD include:

1.  Genotype calling: ANGSD can estimate genotype likelihoods from sequencing data without actually calling genotypes, which helps reduce biases and errors introduced by hard genotype calls.
2.  SNP discovery: The software can identify Single Nucleotide Polymorphisms (SNPs) and estimate their allele frequencies while accounting for sequencing errors and varying levels of sequencing depth.
3.  Genetic association studies: ANGSD can perform Genome-Wide Association Studies (GWAS) and estimate genotype-phenotype associations using mixed linear models or generalized linear models.
4.  Population structure: ANGSD can estimate population structure and admixture proportions using principal component analysis (PCA) or model-based approaches.
5.  Selection scans: The software can detect signals of positive or balancing selection, using various statistics like FST, Tajima's D, and nucleotide diversity.
6.  Handling of various file formats: ANGSD can work with various input file formats, including BAM, CRAM, and VCF files, and output results in multiple formats suitable for downstream analyses.

Overall, ANGSD is a versatile and powerful tool for analyzing NGS data, particularly for non-model organisms or low-coverage sequencing projects. It provides a comprehensive suite of analysis tools that cater to various research objectives in population genetics and genomics.

#### a) Run heterozygosity and realSFS with ANGSD using the bash script below

Because of our time and resource limitations, we are running this analysis only for 2 short chromosomes: chr 11 and chr 12.
This script will loop through chromosomes 11 to 12, running the ANGSD command and then the realSFS command for each chromosome.
ANGSD will calculate the site allele frequency (SAF) likelihood based on individual genotype likelihoods assuming HWE. It generates 3 output binary files (angsdput.saf.idx, angsdput.saf.pos.gz and angsdput.saf.gz).
realSFS will run an optimization of the .saf file, which will estimate the Site Frequency Spectrum (SFS). We are estimating it in windows of a fixed number of sites (200,000 bp). It generates an output file .est.ml, which we will further manipulate to plot the results.

- Create a new file with a name that ends with `.sh` using vim/nano (e.g. `02_run_genome_wide_het.sh`),
- Copy the script below to the file,
- Replace the `<YOUR_USERNAME>` with your Anvil username, 
- Run it on Anvil using `sbatch <job_name>.sh`, and
- Check the job status with: `squeue -u <your_username>`.

```bash
#!/bin/bash
#SBATCH --job-name genome_wide_het
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=1        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=0:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/genome_wide_het.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/genome_wide_het.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

#Load angsd
module load biocontainers/default
module load angsd/0.940

#Choose individual
SAMPLE="CH0893" #CH0893 and DS1507

#Choose the output directory
output_directory="/anvil/scratch/YOUR_USERNAME/05_genetic_diversity"

#Set variables. Do not change this part.
input_bam_file="/anvil/projects/x-bio240351/shared_data/bams/${SAMPLE}.loxAfr4_NC000934"
ancestral_fasta_file="/anvil/projects/x-bio240351/shared_data/reference/loxAfr4.fasta"
reference_fasta_file="/anvil/projects/x-bio240351/shared_data/reference/loxAfr4.fasta"

#Loop through scaffolds 11 to 12

for chr in {11..12}; do

    #Run ANGSD
    angsd -P 1 -i ${input_bam_file}_${chr}.bam -anc ${ancestral_fasta_file} -dosaf 1 -gl 1 -C 50 -minQ 20 -minmapq 30 -out ${output_directory}/$SAMPLE.chr_${chr} -ref ${reference_fasta_file}

    #Run realSFS
    realSFS -nsites 200000 ${output_directory}/$SAMPLE.chr_${chr}.saf.idx -fold 1 > ${output_directory}/$SAMPLE.chr_${chr}.est.ml

done
```

>[!NOTE]
>This script run ANGSD for a single individual defined in the variable `$SAMPLE`. Adjust the script to run it for a second sample named `DS1507`.

Make sure you understand each option in the ANGSD command line:

-   `P <threads>` \- Sets the number of threads to be used in parallel.
-   `i <input_bam_file>` \- Specifies the input file as a BAM file.
-   `anc <ancestral_fasta_file>` \- Specifies the ancestral fasta reference file.
-   `dosaf <dosaf_value>` \- Computes the Site Frequency Spectrum (SFS) based on the genotype likelihoods.
-   `gl <genotype_likelihood_method>` \- Specifies the method used for calculating genotype likelihoods.
-   `C <base_quality_adjustment>` \- Adjusts the base quality score by a specified value before using it.
-   `minQ <min_base_quality>` \- Sets the minimum base quality score required.
-   `minmapq <min_mapping_quality>` \- Sets the minimum mapping quality score required.
-   `fold <fold_value>` \- Indicates whether you are analyzing folded SFS or unfolded SFS.
-   `out <output_file>` \- Specifies the output file path and name.
-   `ref <reference_fasta_file>` \- Specifies the reference fasta file.
-   `r <region_of_interest>` \- Specifies the region of interest. ANGSD does not differentiate the chromosomes if you run the whole genome at once, that is why we need to use the ‘region’ variable when running ANGSD to specify the chromosome/scaffold. This will be useful to look the heterozygosity throughout the genome.

Also, make sure you understand each option in the realSFS command line:

-   `nsites <number_of_sites>` \- Specifies the number of sites to be considered for the estimation. 
-   `<input_saf_idx_file>` \- Specifies the input .saf.idx file, which is the output from the ANGSD program that contains information about the SFS.
-   `> <output_est_ml_file>` \- Specifies the output file path and name for the maximum likelihood estimate of the SFS.

#### b) Prepare files and plot heterozygosity along the scaffold

Now let's prepare the `est.ml` files for plotting. The script below will combine the results for the two samples, and add the sample name and scaffold number for each line in our output. This will make our work easier when we want to plot our results.
Copy this script to the server, replace `YOUR_USERNAME` by your Anvil userneme and run it using `sbatch`.

```bash
#!/bin/bash
#SBATCH --job-name prep_genome_wide_het
#SBATCH -A bio240351  # Allocation name
#SBATCH --nodes=1         # Total # of nodes (must be 1 for serial job)
#SBATCH --ntasks=100M        # Total # of MPI tasks (should be 1 for serial job)
#SBATCH --time=0:30:00    # Total run time limit (hh:mm:ss)
#SBATCH -o /anvil/scratch/YOUR_USERNAME/log/prep_genome_wide_het.o%j      # Name of stdout output file
#SBATCH -e /anvil/scratch/YOUR_USERNAME/log/prep_genome_wide_het.e%j      # Name of stderr error file
#SBATCH -p wholenode  # Queue (partition) name

# Set the input and output directories
output_directory="/anvil/scratch/YOUR_USERNAME/05_genetic_diversity"

# Define the list of sample IDs (space-separated)
SAMPLES="CH0893 DS1507"

# Create or clear the final concatenated output file
output_file="${output_directory}/genome_wide_het_concatenated.txt" > "${output_file}"

# Loop through the sample IDs
for SAMPLE in $SAMPLES; do
    # Loop through scaffolds 11 to 12
    for i in {11..12}; do
        # Define the current file path
        input_file="${output_directory}/${SAMPLE}.chr_${i}.est.ml"

        # Check if the input file exists
        if [ -f "${input_file}" ]; then
            # Calculate the number of lines in the current file
            num_lines=$(wc -l < "${input_file}")

            # Add the number of lines, sample name, and scaffold number to each line of the output file
            awk -v lines="${num_lines}" -v sample="${SAMPLE}" -v scaffold="${i}" '{print lines, sample, "chr_" scaffold, $0}' "${input_file}" > "${output_directory}/${SAMPLE}.chr_${i}.est.ml.annotated"

            # Move the annotated file to the original file
            mv "${output_directory}/${SAMPLE}.chr_${i}.est.ml.annotated" "$input_file"

            # Concatenate the annotated file to the final output file
            cat "${input_file}" >> "${output_file}"
        else
            echo "Warning: File ${input_file} not found, skipping." >&2
        fi
    done
done
```

#### c) Plot the results for the two chromosomes and two individuals using R:

- Load the R module by entering `module load r/4.1.0` in the terminal.
- Start an R session by typing `R`.
- Copy and paste the commands in the script below in the terminal. 
- Use `quit()` to close R.
- Use Anvil browser to visualize the output file `genome_wide_het.png`.


```r
# Load required libraries
library(dplyr)
library(tidyverse) 
library(viridis)   
library(scales)   

# Read the data
het_master <- read.table("genome_wide_het_concatenated.txt")

# Data manipulation
het_master <- het_master %>%
  rename(sample = V2, chromosome = V3) %>%
  mutate(heterozygosity = V5 / (V4 + V5)) %>%   # Calculate heterozygosity as V5 / (V4 + V5)
  group_by(sample, chromosome) %>% 
  mutate(position = row_number()) %>%
  ungroup()                             

# Plot all chromosomes
barplot <- ggplot(het_master, aes(x = position, y = heterozygosity)) + 
  geom_bar(stat = "identity", aes(fill = factor(chromosome)), width = 1) + 
  scale_fill_viridis(discrete = TRUE) +
  facet_grid(chromosome ~ sample, scales = "free_x") + 
  labs(x = NULL, y = "Heterozygosity\n") + 
  scale_y_continuous(labels = comma) + 
  scale_x_continuous(labels = comma) + 
  theme_minimal() + 
  theme(legend.position = "none", 
    strip.text.x = element_text(face = "bold"),
    strip.text.y = element_text(face = "bold"),
    panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(), 
    panel.spacing.x = unit(0, "line"),  panel.border = element_rect(color = "black", fill = NA, size = 0.25))

print(barplot)
ggsave("genome_wide_het.png", units = "in", width = 10, height = 5)
```

> [!IMPORTANT]
> :elephant::grey_question: How is heterozygosity distributed along the scaffold? Is the distribution consistent across the scaffold, or are there areas with notably higher or lower heterozygosity?? Are there gaps or regions with no detectable heterozygosity?
